internal/rbd/group_controller.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util/log"
+)
+
+// CreateVolumeGroupSnapshot receives a list of volume handles and is requested
+// to create a list of snapshots that are created at the same time. This is
+// similar (although not exactly the same) to consistency groups.
+//
+// RBD has a limitation that an image can only belong to a single group. It is
+// therefore required to create a temporary group, add all images, create the
+// group snapshot and remove all images from the group again. This leaves the
+// group and its snapshot around, the group snapshot can be inspected to list
+// the snapshots of the images.
+func (cs *ControllerServer) CreateVolumeGroupSnapshot(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+) (*csi.CreateVolumeGroupSnapshotResponse, error) {
+	var (
+		err           error
+		vg            types.VolumeGroup
+		groupSnapshot types.VolumeGroupSnapshot
+
+		// the VG and VGS should not have the same name
+		vgName  = req.GetName() + "-vg" // stable temporary name
```

Comment:
What happens if there is already a volumeGroup that was created for VolumeGroup that was used for replication?

Reply:
Because an RBD-image can only belong to one RBD-group at the time, a volume can either be replicated as a group (long living group), or snapshotted as a group (short living group), not both.

It is possible that a replication-group and snapshot-group contain the same source volumes. In that case, there would be no need to create (and delete) the group for snapshotting. This is something that can be added in the future.